src/telemetry/exporters.ts

```diff
@@ -35,39 +35,45 @@ export class TelemetryExporter implements ITelemetryExporter {
   }
 
   async export(signals: TelemetrySignal[]): Promise<void> {
-    return await new Promise<void>((resolve) => {
-      // Sort out traces and logs
-      const exportSpans: ReadableSpan[] = [];
-      const exportLogs: ReadableLogRecord[] = [];
-      signals.forEach((signal) => {
-        if (isTraceSignal(signal)) {
-          exportSpans.push(signal as ReadableSpan);
-        }
-        if (isLogSignal(signal)) {
-          exportLogs.push(signal as ReadableLogRecord);
-        }
-      });
-
-      if (exportSpans.length > 0 && this.tracesExporter) {
-        this.tracesExporter.export(exportSpans, (results: ExportResult) => {
-          if (results.code !== ExportResultCode.SUCCESS) {
-            console.warn(`Trace export failed: ${results.code}`);
-            console.warn(results);
-          }
-        });
+    // Sort out traces and logs
+    const exportSpans: ReadableSpan[] = [];
+    const exportLogs: ReadableLogRecord[] = [];
+    signals.forEach((signal) => {
+      if (isTraceSignal(signal)) {
+        exportSpans.push(signal as ReadableSpan);
       }
-
-      if (exportLogs.length > 0 && this.logsExporter) {
-        this.logsExporter.export(exportLogs, (results: ExportResult) => {
-          if (results.code !== ExportResultCode.SUCCESS) {
-            console.warn(`Log export failed: ${results.code}`);
-            console.warn(results);
-          }
-        });
+      if (isLogSignal(signal)) {
+        exportLogs.push(signal as ReadableLogRecord);
       }
-
-      resolve();
     });
+    const tasks : Promise<void>[] = [];
+    if (exportSpans.length > 0 && this.tracesExporter) {
+      tasks.push(
+        new Promise<void>((resolve) => {
+          this.tracesExporter?.export(exportSpans, (results: ExportResult) => {
+            if (results.code !== ExportResultCode.SUCCESS) {
+              console.warn(`Trace export failed: ${results.code}`);
+              console.warn(results);
+            }
+            resolve();
+          });
+        })
```

Comment:
I think this "looks" more complicated than it needs to be. My personal preference would be to do something like:

```
const callback = (results: ExportResult) => {
    if (results.code !== ExportResultCode.SUCCESS) {
      console.warn(`Trace export failed: ${results.code}`);
      console.warn(results);
    }
};

this.tracesExporter?.export(exportSpans, callback);

tasks.push(callback);
```

Reply:
@maxdml there's an extra trick here because `callback` needs to call the `resolve` that's declared in the promise. So the callback needs to be declared within the promise, otherwise `resolve` is out of scope.

So maybe like this:
```
      const exportTask = new Promise<void>((resolve) => {
        const exportCallback = (results: ExportResult) => {
          if (results.code !== ExportResultCode.SUCCESS) {
            console.warn(`Trace export failed: ${results.code}`);
            console.warn(results);
          }
          resolve();
        };
        this.tracesExporter?.export(exportSpans, exportCallback);
      });
      tasks.push(exportTask);
```

... gets rid of one indent level....

Reply:
I see, we do have to pass a resolver to the callback such that, when the otel-js framework calls it, we can get notified. Probably not worth the refactor then (though I like your version above better), but worth a comment to explain what's going on